library/Amazon/SNS/Verify/Validate.hs

```diff
@@ -79,22 +79,27 @@ retrieveCertificate
   => SNSPayload
   -> ExceptT SNSNotificationValidationError m SignedCertificate
 retrieveCertificate SNSPayload {..} = do
-  certUrlStr <- unTryE id $ validateCertUrl snsSigningCertURL
+  certUrlStr <- validateCertUrl snsSigningCertURL
   response <- httpLbs $ parseRequest_ certUrlStr
   pems <- unTryE BadPem $ pemParseLBS $ getResponseBody response
   cert <-
     fromMaybeM (throwE $ BadPem "Empty List") $ pemContent <$> headMay pems
   unTryE BadCert $ decodeSignedCertificate cert
 
-validateCertUrl :: Text -> Either SNSNotificationValidationError String
+validateCertUrl
+  :: Monad m => Text -> ExceptT SNSNotificationValidationError m String
 validateCertUrl certUrl = do
-  uri <- fromMaybeM (Left $ BadUri certUrlStr) $ parseURI certUrlStr
-  if uriScheme uri
-    == validScheme
-    && maybe "" uriRegName (uriAuthority uri)
-    =~ validRegPattern
-    then Right certUrlStr
-    else Left $ BadUri certUrlStr
+  uri <- fromMaybeM (throwE $ BadUri certUrlStr) $ parseURI certUrlStr
+
+  let
+    scheme = uriScheme uri
+    mDomain = uriRegName <$> uriAuthority uri
+
+  unless (scheme == validScheme) $ throwE $ BadScheme scheme
+  domain <- fromMaybeM (throwE $ NoAuthority certUrlStr) mDomain
```

Comment:
I appreciate being explicit about this and no longer just tacitly relying on the hope that `""` will not satisfy `validRegPattern`